### ISSUE TITLE
[BUG] memory related test failures post CI simplification PR

### ISSUE THREAD

**Issue by fkiraly** [alan-turing-institute/sktime#3103]
Since we merged the "CI simplification" PR https://github.com/alan-turing-institute/sktime/pull/2896, I have observed a number of mem-out failures of tests, in the OS/version combinations that the CI simplification has addressed.

I cannot claim a causal relation, but the timing makes me suspicious.

Examples of mem-outs:
https://github.com/alan-turing-institute/sktime/runs/7524247553
https://github.com/alan-turing-institute/sktime/runs/7522495684

Any good ideas, @lmmentel?
Maybe @freddyaboulton, @RNKuhns have an idea?

**Comment by fkiraly** [comment]
FYI, #3095 is purely doc related (code of conduct).

While the proposed changes might be contentious to some, it is hard to imagine any plausible technical reason for why the changes would lead to failure in the classification module...

**Comment by fkiraly** [comment]
One more, also on mac: https://github.com/alan-turing-institute/sktime/runs/7523631638

All classification related, maybe that's another possible lead? @MatthewMiddlehurst ?

**Comment by lmmentel** [comment]
Mac failures might be related to macos-10.15 being deprecated https://github.com/actions/virtual-environments/issues/5583 I'll make a PR to bump the OS version of the CI host.

**Comment by MatthewMiddlehurst** [comment]
This seems to be affecting #3090 as well for a range of classifiers and transformers. Most of the failures have been untouched for quite a while as far as I am aware.

**Comment by fkiraly** [comment]
I have an explanation for 14 workers, @lmmentel: they are not concurrent, but sequential. One after another, the workers crash, so `pytest-xdist` spins up new ones in an attempt to try again.

**Comment by fkiraly** [comment]
Revert of recent CI changes seems to fix the issues, see #3118.

The recent changes that are reverted consist of three things:

* testing notebooks on py 3.9, not 3.7, by @fkiraly 
* codecov-action changed from version 2 to version 3, by @miraep8 
* merging separate linux and mac CI elements into a single matrix, by @lmmentel

My suspicion is towards the merge of linux/max since it appears as the most substantial change, but we cannot rule anything out at the moment imo.

**Comment by fkiraly** [comment]
(@lmmentel, @miraep8, not trying to play the blame game here - after all I've approved the PR, and all looked very sensible to me too!)

**Comment by miraep8** [comment]
I am definitely in favor of rolling back the changes!  I think in my case at least (ie updating codecov from 2 to 3) I can't claim to fully understand what that changed.  The hope was - not much....; but seems not to have been the case 🙃   Thanks for investigating!

**Comment by fkiraly** [comment]
Ok, new data point:
I restarted https://github.com/alan-turing-institute/sktime/pull/3117
(0.12.1 state) and it *also* failed on mac.

That would be strong evidence in favour of the "sth is happening to mac on GitHub cloud" hypothesis

**Comment by fkiraly** [comment]
>The hope was - not much....; but seems not to have been the case

@miraep8, we cannot infer that - the failures are statistical (so evidence increases/decreases with re-runs), and even if the changes were responsible, there were three in total, of which I think yours is less plausibly the cause.

**Comment by fkiraly** [comment]
@RavenRudi, while working on #3126 and looking at #767, I noticed that a while ago you introduced parallelization to the `RocketClassifier`. That appears frequently in some of the failures. Any ideas, in case you still remember? Maybe parallelization related?

**Comment by fkiraly** [comment]
could it be two separate things, @lmmentel?

mac upgrade *plus* something in merging the two items?

That would explain why the reverts and historical states fail only on mac?

**Comment by fkiraly** [comment]
The following seems to work - removing these estimators:

* `HIVECOTEV1`
* `HIVECOTEV2`
* `RandomIntervalSpectralEnsemble`
* `RandomInvervals`
* `RandomIntervalSegmenter`
* `RandomIntervalFeatureExtractor`
* `RandomIntervalClassifier`
* `MiniRocket`
* `MatrixProfileTransformer`

**Comment by fkiraly** [comment]
`MiniRocket`:
```python
import multiprocessing
from numba import get_num_threads, njit, prange, set_num_threads, vectorize
```

`RandomIntervalStuff`:
```python
from joblib import Parallel, delayed
from numba import int64, jit, prange
```

`HIVECOTE1` imports and uses `RandomIntervalStuff`

**Comment by fkiraly** [comment]
From the above, I conclude that we must be looking at an interaction of HIVE-COTE, random interval estimators with the Mac set-up.

Why:
* reverting to earlier repository states does not prevent this
* removal/presence is a reliable on/off condition, see https://github.com/alan-turing-institute/sktime/pull/3134

**Comment by MatthewMiddlehurst** [comment]
Note for HIVE-COTE, as it seems all of the individual classifiers are not failing, the issue is most likely in the class itself or the _get_train_probs methods. 

I don't believe it uses any of the RandomIntervalStuff transformers, either. RISE, DrCIF etc. predate these transformers I think, so don't use them currently.

**Comment by MatthewMiddlehurst** [comment]
Possibly a numba issue? Most of them use it in some capacity, not all though, MatrixProfileTransformer uses stumpy (which uses numba https://stumpy.readthedocs.io/en/latest/Tutorial_STUMPY_Basics.html).

Very strange, only thing they seem to have in common algorithmically is that they extract subseries or window through series. Would be a bit of a stretch to say it is something like numpy slicing though.

RandomIntervalSpectralEnsemble seems to appear more often than others, same with HC1 which uses RISE. Is there something it is doing in particular which makes it crash so much?

**Comment by fkiraly** [comment]
> I don't believe HC uses any of the RandomInterval transformers, either. RISE, DrCIF etc. predate these I think, so don't use them currently.

Yes, I first suspected it is one of the components which has a domino effect, but that can't be it if you look at precisely what the components are (there is not one component that fails and also sits in the ensembles).

**Comment by fkiraly** [comment]
>Possibly a numba issue?

Also a suspicion I had, but not all failures use `numba`, and not all estimators that use `numba` are failing.

**Comment by fkiraly** [comment]
@haskarb noticed one commonality between the estimators: use of `numba`

**Comment by fkiraly** [comment]
Possibly related to this: https://github.com/sktime/sktime/issues/4266
